tests/unit_tests/steps/test_compile_c.py

```diff
@@ -50,110 +59,138 @@ def test_compile_c_wrong_compiler(content):
 
     # Now check that _compile_file detects the incorrect class of the
     # C compiler
-    mp_common_args = mock.Mock(config=config)
-    with pytest.raises(RuntimeError) as err:
-        _compile_file((None, mp_common_args))
-    assert ("Unexpected tool 'mock_c_compiler' of category "
-            "'FORTRAN_COMPILER' instead of CCompiler" in str(err.value))
+    mp_common_args = Mock(config=config)
+    with raises(RuntimeError) as err:
+        _compile_file((Mock(), mp_common_args))
+    assert str(err.value) == "Unexpected tool 'some C compiler' of category " \
+                             "'FORTRAN_COMPILER' instead of CCompiler"
 
 
 # This is more of an integration test than a unit test
 class TestCompileC:
     '''Test various functionalities of the C compilation step.'''
 
-    def test_vanilla(self, content):
-        '''Ensure the command is formed correctly.'''
-        config, _, expect_hash = content
-        compiler = config.tool_box[Category.C_COMPILER]
-        # run the step
-        with mock.patch("fab.steps.compile_c.send_metric") as send_metric:
-            with mock.patch('pathlib.Path.mkdir'):
-                with mock.patch.dict(os.environ, {'CFLAGS': '-Denv_flag'}), \
-                     pytest.warns(UserWarning, match="_metric_send_conn not set, "
-                                                     "cannot send metrics"):
-                    compile_c(config=config,
-                              path_flags=[AddFlags(match='$source/*',
-                                                   flags=['-I', 'foo/include', '-Dhello'])])
-
-        # ensure it made the correct command-line call from the child process
-        compiler.run.assert_called_with(
-            cwd=Path(config.source_root),
-            additional_parameters=['-c', '-Denv_flag', '-I', 'foo/include',
-                                   '-Dhello', 'foo.c',
-                                   '-o', str(config.prebuild_folder /
-                                             f'foo.{expect_hash:x}.o')],
-        )
-
-        # ensure it sent a metric from the child process
-        send_metric.assert_called_once()
+    def test_vanilla(self, content,
+                     fake_process: FakeProcess, monkeypatch) -> None:
+        """
+        Tests correct use of compiler.
+        """
+        config, _ = content
+
+        monkeypatch.setenv('CFLAGS', '-Denv_flag')
```

Comment:
Oh, I didn't know that - that's neat :)

Reply:
Like anything related to mocking and patching, it has to be used with care, but it can be incredibly useful.